db/internal_stats.cc

```diff
@@ -1265,6 +1273,17 @@ bool InternalStats::HandleNumRunningCompactions(uint64_t* value, DBImpl* db,
   return true;
 }
 
+bool InternalStats::HandleNumRunningCompactionSortedRuns(uint64_t* value,
+                                                         DBImpl* db,
+                                                         Version* /*version*/) {
+  uint64_t sorted_runs = 0;
+  for (auto* cfd : *db->versions_->GetColumnFamilySet()) {
```

Comment:
Is the DB mutex held when this function is called? The description of `ColumnFamlySet` says the mutex must be held - https://github.com/facebook/rocksdb/blob/1341c0c670a8134909a797587b97cfe4d5bf6645/db/column_family.h#L709.

Reply:
Good catch.

I don't think the DB mutex is currently being held when this function is called, so I updated the method to acquire it

https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl.cc#L1166-L1191 shows the mutex being acquired to iterate over the column family set

But in https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl.cc#L1193-L1194 the mutex is not held during the call to `GetPropertyInfo`.

Reply:
The mutex was acquired here https://github.com/facebook/rocksdb/blob/a10b4aa9a38571d992b5588a2f08369bc1e19b18/db/db_impl/db_impl.cc#L1166 for GetPropertyInfo

Reply:
> I don't think the DB mutex is currently being held when this function is called

I'm confused. I thought mutex was already acquired before calling HandleNumRunningCompactionSortedRuns() - for example, see DBImpl::DumpStats() . 

Then why did we need to acquire again `InstrumentedMutexLock l(db->mutex());` in this HandleNumRunningCompactionSortedRuns?? 

Can we just assert `db_mutex->AssertHeld();`??

Reply:
> for example, see DBImpl::DumpStats() 

there were too many curly braces in https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl.cc#L1166-L1189. I thought I saw a closing curly brace that put `InstrumentedMutexLock l(&mutex_);` out of scope

Reply:
@anand1976 it seems like we don't need to acquire the mutex inside this handler function. However @hx235 I put out https://github.com/facebook/rocksdb/pull/13367 yesterday, which I think is likely a cleaner solution. Feel free to comment/take a look.

Reply:
If there is no locking issue imposed in this PR, what https://github.com/facebook/rocksdb/pull/13367 is trying to addressed that this PR couldn't do cleanly??

Reply:
I mean it is subjective which approach is "cleaner" but some of the nice aspects of #13367 compared to this PR are:
1. We track the statistic as a DB-wide statistic, rather than tracking at the column family level and then aggregating at the end
2. Stats gathering occurs at "natural" places where we already are trying to fire off event listeners. Contrast that with this PR, where we record stats in the constructor/destructor, which feels strange to me.
3. Even though we think that we won't run into an issue with the db mutex when reporting this stat, it is nice that in #13367  locking/deadlocks is not a concern to begin with